### PR TITLE
Add http-request-function and processor app

### LINKED ISSUE #36: Migrate httpclient processor apps
Convert the current httpclient processor as a function and then generate apps for the binders.

### PR DESCRIPTION
**Pull request by dturanski**
Resolves #36

### CODE REVIEW

**Comment by sobychacko** [comment]
@dtur Could you also add a simple `README` on the functions side? See [this PR](https://github.com/spring-cloud/stream-applications/pull/47) where I tried to add some basic info. All it needs is some information about how one may inject it as a standalone function and use it. We can also link to the properties class there.

File: applications/processor/http-request-processor/src/test/resources/META-INF/spring.binders [hunk 1]
```diff
@@ -0,0 +1,2 @@
+integration:\
+org.springframework.cloud.stream.binder.test.TestChannelBinderConfiguration
```
Comment (artembilan) [inline applications/processor/http-request-processor/src/test/resources/META-INF/spring.binders:2]: New line in the end of file.

File: applications/processor/http-request-processor/src/main/resources/META-INF/dataflow-configuration-metadata-whitelist.properties [hunk 2]
```diff
@@ -0,0 +1,2 @@
+configuration-properties.classes=org.springframework.cloud.stream.app.processor.http.request.HttpRequestProcessorProperties,\
+  org.springframework.cloud.stream.app.processor.http.request.HttpRequestProcessorProperties$Retry
```
Comment (artembilan) [inline applications/processor/http-request-processor/src/main/resources/META-INF/dataflow-configuration-metadata-whitelist.properties:2]: Looks like there is no `Retry` class any more.

File: applications/processor/http-request-processor/src/main/java/org/springframework/cloud/stream/app/processor/http/request/HttpRequestProcessorProperties.java [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+ * @author Waldemar Hummer
+ * @author Mark Fisher
+ * @author Christian Tzolov
+ * @author Artem Bilan
+ * @author David Turanski
+ */
+
+@Validated
+@ConfigurationProperties("http.request.processor")
+public class HttpRequestProcessorProperties implements HttpRequestProperties {
```
Comment (artembilan) [inline applications/processor/http-request-processor/src/main/java/org/springframework/cloud/stream/app/processor/http/request/HttpRequestProcessorProperties.java:46]: Even though you don't create a function as a bean over here, you still can configure `http.request` properties.
So, I don't see anything wrong to have two these configuration classes on the processor and do respective logic to distribute properties combination. This way we indeed won't need a common interface to support and will avoid some confusion in the target projects.

File: applications/processor/http-request-processor/src/main/resources/META-INF/dataflow-configuration-metadata-whitelist.properties (hunk 2 above)
Comment (dturanski) [reply applications/processor/http-request-processor/src/main/resources/META-INF/dataflow-configuration-metadata-whitelist.properties:2]: Yes. We discussed that. It no longer applies here. The binder already implements Retry.

File: applications/processor/http-request-processor/src/main/java/org/springframework/cloud/stream/app/processor/http/request/HttpRequestProcessorProperties.java (hunk 3 above)
Comment (dturanski) [reply applications/processor/http-request-processor/src/main/java/org/springframework/cloud/stream/app/processor/http/request/HttpRequestProcessorProperties.java:46]: This is a consequence of the boot metadata/doc plugin. The only way to generate all the properties in the README AFAICT.  I tried some other (obvious) combinations.

Comment (dturanski) [reply applications/processor/http-request-processor/src/main/java/org/springframework/cloud/stream/app/processor/http/request/HttpRequestProcessorProperties.java:46]: Also:
* The validation rules (e.g., only one of urlExpression, url) require the processor properties to "inherit" or compose with the function properties.  The validation rules are different for function properties (e.g. url is required).
* The same prefix for all properties in the app context is less confusing IMHO

File: applications/processor/http-request-processor/src/main/resources/META-INF/dataflow-configuration-metadata-whitelist.properties (hunk 2 above)
Comment (artembilan) [reply applications/processor/http-request-processor/src/main/resources/META-INF/dataflow-configuration-metadata-whitelist.properties:2]: So, are you going to fix this then?

File: applications/processor/http-request-processor/src/main/java/org/springframework/cloud/stream/app/processor/http/request/HttpRequestProcessorProperties.java (hunk 3 above)
Comment (artembilan) [reply applications/processor/http-request-processor/src/main/java/org/springframework/cloud/stream/app/processor/http/request/HttpRequestProcessorProperties.java:46]: OK. Thanks. Resolved then!

File: applications/processor/http-request-processor/src/main/resources/META-INF/dataflow-configuration-metadata-whitelist.properties [hunk 4]
```diff
@@ -0,0 +1,2 @@
+configuration-properties.classes=org.springframework.cloud.stream.app.processor.http.request.HttpRequestProcessorProperties,\
```
Comment (artembilan) [inline applications/processor/http-request-processor/src/main/resources/META-INF/dataflow-configuration-metadata-whitelist.properties:1]:
```suggestion
configuration-properties.classes=org.springframework.cloud.stream.app.processor.http.request.HttpRequestProcessorProperties
```

File: applications/processor/http-request-processor/src/main/resources/META-INF/dataflow-configuration-metadata-whitelist.properties (hunk 2 above)
Comment (artembilan) [reply applications/processor/http-request-processor/src/main/resources/META-INF/dataflow-configuration-metadata-whitelist.properties:2]:
```suggestion
```

**Pull request merged**